server/sonar-server/src/main/java/org/sonar/server/ui/ContextImpl.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.ui;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.api.web.page.Context;
+import org.sonar.api.web.page.Page;
+
+class ContextImpl implements Context {
+  private static final Logger LOGGER = Loggers.get(ContextImpl.class);
+
+  private Map<String, Page> pagesByPath = new HashMap<>();
+
+  @Override
+  public Context addPage(Page page) {
+    Page existingPageWithSameKey = pagesByPath.putIfAbsent(page.getKey(), page);
+    if (existingPageWithSameKey != null) {
+      LOGGER.warn("Page '{}' cannot be loaded. Another page with key '{}' already exists.", page.getName(), page.getKey());
```

Comment:
it will be at server startup. Do you really want to fail the server startup because 2 pages have the same key?